heat/nn/data_parallel.py

```diff
@@ -0,0 +1,275 @@
+            def __init__(self):
+                super(TestingModel, self).__init__()
+                self.net1 = ht.nn.Linear(10, 10)
+                self.relu = ht.nn.ReLU()
+                self.net2 = ht.nn.Linear(10, 5)
+
+            def forward(self, x):
+                return self.net2(self.relu(self.net1(x)))
+
+        ht_model = ht.nn.DataParallel(TestingModel(), comm, optimizer)
+
+    and a requirement of giving a HeAT communicator (``comm``) as well as a Torch optimizer (``optimizer``). For the
+    given model both the ``__init__()`` and ``forward()`` functions must be defined in the class defining the network.
+
+    It is highly recommended that a HeAT DataLoader is used, see :func:`..utils.data.datatools.DataLoader`. The
+    default communications scheme for this is blocking. The blocking scheme will average the model parameters during
+    the backwards step, synchronizing them before the next model iteration.
+
+    To use the non-blocking communications for parameter updates, negate the optional flag ``blocking``.
+
+    Attributes
+    ----------
+    module : torch.nn.Module
+        the local module
+    comm : heat.MPICommunicator
+        HeAT communicator to use
+    optimizer : torch.optim.Optimizer
+        Optimizer used for parameter updates.
+    blocking : bool (optional)
+        Flag for blocking synchronization. If not given, synchronization is blocking by default.
+    """
+
+    def __init__(
+        self,
+        module: torch.nn.Module,
+        comm: ht.MPICommunication,
+        optimizer: torch.optim.Optimizer,
+        blocking: bool = True,
+    ):
+        super(DataParallel, self).__init__()
+        self.module = module
+        self.comm = comm
+        self.optimizer = optimizer
+        self.blocking = blocking
+
+        self._layer_wait_handles = OrderedDict()
+        self._fwd_hook_handles = list()
+        # set of layers' names with active wait handles (only relevant for non-blocking)
+        self._active_layers = set()
+        # slices of parameters belonging to one and the same layer
+        self._param_slices = dict()
+        # pytorch internal parameter indexing
+        self._param_indices = dict()
+        # reference of optimizer's params
+        self._params_ref = None
+        # flag indicating if optimizer should take a step during next iteration (only relevant for non-blocking)
+        self._update_next = False
+
+        # check if optimizer matches module
+        if list(module.parameters()) != optimizer.param_groups[0]["params"]:
+            raise ValueError("given module and optimizer don't share same parameters.")
+        else:
+            # take reference of optimizer's params
+            self._params_ref = optimizer.param_groups[0]["params"]
+
+        # get parameter indexing and slices
+        start_idx = 0
+        layer_name_prev = None
+        for idx, (name, param) in enumerate(module.named_parameters()):
```

Comment:
i believe that all of the layers must have names. this is done automatically in the background by torch. but @lehr-fa would know more

Reply:
I don't really want to do this now. It was rather meant to give the possibility to pass a subset. You can disable parameters with, requires_grad, but it is also allowed to pass only a part of the parameters, simple example: pass  filter(lambda p: p.requires_grad, model.parameters()) instead of model.parameters() after some parameters are marked inactive

Reply:
In case of  the Sequential class, the layers are numbered. In case of the Module class, layers have the names you assign to them in the init() method. e.g. 
self.fc = nn.Linear(D, H)
names will be:
fc.weight
fc.bias

Reply:
I needed an unique attribute to distinguish between layers and get a clear assignment. Naming works as @krajsek explained. In both cases, there is no possibility that two different layers have the same name. But I see a little mistake here, occurring when one creates nested networks: The string returned by `named_parameters()` may only be split at the rightmost occurrence of dot delimiter, currently it is split at the leftmost occurrence.